File: mobile/modules/lorca-core/ios/ShimmerView.swift
import ExpoModulesCore

/// A container whose contents shimmer, as the Mac's working row does: a gradient mask dims
/// them to 40% and sweeps a bright band, 30% of the width, from the leading edge to the
/// trailing one in 1.5 s, resting a quarter second between sweeps.
public class ShimmerViewModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ShimmerView")
    View(ShimmerView.self) {}
  }
}

final class ShimmerView: ExpoView {
  /// The children live in a view of this one's own: React Native clears the mask of the view
  /// it styles every time it restyles it.
  private let content = UIView()
  private let shimmer = ShimmerLayer()

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    content.layer.mask = shimmer
    addSubview(content)
  }

  override func mountChildComponentView(_ childComponentView: UIView, index: Int) {
    content.insertSubview(childComponentView, at: index)
  }

  override func unmountChildComponentView(_ childComponentView: UIView, index: Int) {
    childComponentView.removeFromSuperview()
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    content.frame = bounds
    shimmer.frame = content.bounds
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()
    window == nil ? shimmer.stop() : shimmer.start()
  }
}

/// The mask. Every sweep keeps time with the media clock, so a row made again mid-sweep (a
/// list re-render) carries on where the last one was instead of starting over.
private final class ShimmerLayer: CAGradientLayer {
  private static let band: CGFloat = 0.3
  private static let sweep: CFTimeInterval = 1.5
  private static let rest: CFTimeInterval = 0.25

  override init() {
    super.init()
    let dim = UIColor.black.withAlphaComponent(0.4).cgColor
    colors = [dim, UIColor.black.cgColor, dim]
    startPoint = CGPoint(x: -Self.band, y: 0.5)
    endPoint = CGPoint(x: 0, y: 0.5)
  }

  override init(layer: Any) {
    super.init(layer: layer)
  }

  @available(*, unavailable)
  required init?(coder: NSCoder) { fatalError() }

  /// The mask follows its view's frame at once.
  override func action(forKey event: String) -> CAAction? { nil }

  func start() {
    guard animation(forKey: "shimmer") == nil else { return }
    let start = CABasicAnimation(keyPath: "startPoint")
    start.fromValue = NSValue(cgPoint: CGPoint(x: -Self.band, y: 0.5))
    start.toValue = NSValue(cgPoint: CGPoint(x: 1, y: 0.5))
    let end = CABasicAnimation(keyPath: "endPoint")
    end.fromValue = NSValue(cgPoint: CGPoint(x: 0, y: 0.5))
    end.toValue = NSValue(cgPoint: CGPoint(x: 1 + Self.band, y: 0.5))
    start.duration = Self.sweep
    end.duration = Self.sweep

    let shimmer = CAAnimationGroup()
    shimmer.animations = [start, end]
    shimmer.duration = Self.sweep + Self.rest
    shimmer.repeatCount = .infinity
    let now = convertTime(CACurrentMediaTime(), from: nil)
    shimmer.beginTime = now - now.truncatingRemainder(dividingBy: shimmer.duration)
    add(shimmer, forKey: "shimmer")
  }

  func stop() {
    removeAnimation(forKey: "shimmer")
  }
}

File: mobile/modules/lorca-core/android/src/main/java/app/lorca/core/ShimmerView.kt
package app.lorca.core

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Shader
import android.os.SystemClock
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.views.ExpoView

/// A container whose contents shimmer, as the Mac's working row does: a gradient mask dims
/// them to 40% and sweeps a bright band, 30% of the width, from the leading edge to the
/// trailing one in 1.5 s, resting a quarter second between sweeps.
class ShimmerViewModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ShimmerView")
    View(ShimmerView::class) {}
  }
}

/// The children draw into a layer that the band then masks. Every sweep keeps time with the
/// uptime clock, so a row made again mid-sweep carries on where the last one was.
class ShimmerView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  private val mask = Paint(Paint.ANTI_ALIAS_FLAG).apply { xfermode = PorterDuffXfermode(PorterDuff.Mode.DST_IN) }
  private val shift = Matrix()
  private var gradientWidth = 0

  override fun dispatchDraw(canvas: Canvas) {
    val w = width.toFloat()
    val h = height.toFloat()
    if (w <= 0f || h <= 0f) {
      super.dispatchDraw(canvas)
      return
    }
    if (gradientWidth != width) {
      mask.shader = LinearGradient(0f, 0f, BAND * w, 0f, intArrayOf(DIM, Color.BLACK, DIM), null, Shader.TileMode.CLAMP)
      gradientWidth = width
    }
    // The band's leading edge runs from 30% of the width before the start to the end over the
    // sweep, then waits off the start for the rest.
    val phase = (SystemClock.uptimeMillis() % PERIOD_MS).toFloat() / SWEEP_MS
    val start = if (phase < 1f) -BAND + phase * (1f + BAND) else -BAND
    shift.setTranslate(start * w, 0f)
    mask.shader.setLocalMatrix(shift)

    val layer = canvas.saveLayer(0f, 0f, w, h, null)
    super.dispatchDraw(canvas)
    canvas.drawRect(0f, 0f, w, h, mask)
    canvas.restoreToCount(layer)
    postInvalidateOnAnimation()
  }

  private companion object {
    const val BAND = 0.3f
    const val SWEEP_MS = 1500L
    const val PERIOD_MS = 1750L
    val DIM = Color.argb(102, 0, 0, 0)
  }
}
